server/tasks/process_file.py

```diff
@@ -116,84 +73,43 @@ def streaming_parent_task(
-    logger.info(f"Submitting task {result_handler_task.name}")
-    output_result = result_handler_task.delay(
-        input_job_id=input_job_id,
-        output_topic_name=output_topic_name,
-        result_handler=result_handler,
-        batch_size=batch_size,
-    )
-    output_job_id = output_result.id
-    logger.info(
-        f"Task {result_handler_task.name} submitted with job_id {output_job_id}"
-    )
-
-    # Store input and output job IDs in parent task metadata
-    # Need to pass state as well otherwise its overwritten to None
-    self.update_state(
-        state=states.STARTED,
-        meta={"input_job_id": input_job_id, "output_job_id": output_job_id},
-    )
-
-    input_job = process_file_streaming.AsyncResult(input_job_id)
-    output_job = process_streaming_output.AsyncResult(output_job_id)
-
-    terminal_statuses = ["SUCCESS", "FAILURE", "REVOKED"]
+    async def run_streaming():
+        input_task_done = asyncio.Event()
+        async with asyncio.TaskGroup() as task_group:
+            task_group.create_task(
+                async_process_streaming_input(input_task_done, agent)
+            )
+            task_group.create_task(
+                async_process_streaming_output(
+                    input_task_done, output_topic_name, result_handler, batch_size
+                )
+            )
 
-    while (
-        input_job.status not in terminal_statuses
-        or output_job.status not in terminal_statuses
-    ):
-        time.sleep(1)
+    asyncio.run(run_streaming())
 
     # clean up kafka topics
     delete_topic(input_topic_name)
     delete_topic(output_topic_name)
 
     logger.info("Both input and output jobs complete")
 
-    # Update parent task status to SUCCESS and pass metadata again
-    # otherwise its overwritten to None
-    self.update_state(
-        state=states.SUCCESS,
-        meta={"input_job_id": input_job_id, "output_job_id": output_job_id},
-    )
-
-    # This makes it so Celery doesnt update the tasks state again, which would wipe out the custom metadata we added
-    # It will retain that state we set above
-    raise Ignore()
-
-
-@app.task(
-    name="process_file_streaming",
-    track_started=True,
-    serializer="pickle",
-    task_time_limit=settings.task_time_limit_sec,
-)
-def process_file_streaming(agent: Agent):
-    # agent's kafka_bootstrap servers and kafka topics should be set in parent task
 
-    # need to keep these in the same event loop
-    async def run_fn():
+async def async_process_streaming_input(input_task_done: asyncio.Event, agent: Agent):
+    try:
         # start up kaka producer and consumer
         await agent.environment.initialize()
         # Run the agent
         await agent.arun()
+        input_task_done.set()
         # shut down kaka producer and consumer
         await agent.environment.finalize()
-
-    asyncio.run(run_fn())
+    except asyncio.CancelledError as e:
+        await agent.environment.finalize()
+        raise e
```

Comment:
is it worth adding a `finally: await agent.environment.finalize()` here? (similar to the finally clause in `async_process_streaming_output`)

Reply:
yup, made this a little more general